File: frontend/open-event-workspace/libs/core/src/lib/search/search.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpParams} from "@angular/common/http";
import {Observable} from "rxjs";
import {
  AccountSearchRequest,
  AccountSearchResponse,
  CategorySearchRequest,
  CategorySearchResponse,
  EventSearchRequest,
  EventSearchResponse
} from "./search-api";
import {BaseService} from "../base-service";

@Injectable({
  providedIn: 'root'
})
export class SearchService extends BaseService {
  constructor(http: HttpClient) {
    super(http, 'search')
    this.retryCount = 0
  }

  setupEvents(): Observable<any> {
    return this.post('setup/event', {})
  }

  setupAccounts(): Observable<any> {
    return this.post('setup/account', {})
  }

  setupCategories(): Observable<any> {
    return this.post('setup/category', {})
  }

  searchEvents(request: EventSearchRequest, page: number, size: number): Observable<EventSearchResponse> {
    return this.search(request, 'event', page, size)
  }

  searchAccounts(request: AccountSearchRequest, page: number, size: number): Observable<AccountSearchResponse> {
    return this.search(request, 'account', page, size)
  }

  searchCategories(request: CategorySearchRequest, page: number, size: number): Observable<CategorySearchResponse> {
    return this.search(request, 'category', page, size)
  }

  private search<R, S>(request: R, suffix: string, page: number, size: number): Observable<S> {
    let params = new HttpParams()
      .set("page", page)
      .set("size", size)
    return this.post(suffix, request, params)
  }
}

File: backend/src/main/kotlin/de/sambalmueslie/openevent/core/search/SearchService.kt
package de.sambalmueslie.openevent.core.search

import de.sambalmueslie.openevent.core.account.api.Account
import de.sambalmueslie.openevent.core.search.account.AccountSearchOperator
import de.sambalmueslie.openevent.core.search.api.*
import de.sambalmueslie.openevent.core.search.category.CategorySearchOperator
import de.sambalmueslie.openevent.core.search.common.SearchOperatorInfo
import de.sambalmueslie.openevent.core.search.event.EventSearchOperator
import io.micronaut.context.annotation.Context
import io.micronaut.data.model.Pageable
import org.slf4j.Logger
import org.slf4j.LoggerFactory

@Context
class SearchService(
    private val eventOperator: EventSearchOperator,
    private val accountOperator: AccountSearchOperator,
    private val categoryOperator: CategorySearchOperator
) {
    companion object {
        private val logger: Logger = LoggerFactory.getLogger(SearchService::class.java)
    }

    private val operators = listOf(eventOperator, accountOperator, categoryOperator).associateBy { it.key() }

    fun searchEvents(actor: Account, request: EventSearchRequest, pageable: Pageable): EventSearchResponse {
        return eventOperator.search(actor, request, pageable)
    }

    fun setupEvents() {
        return eventOperator.setup()
    }

    fun searchAccounts(actor: Account, request: AccountSearchRequest, pageable: Pageable): AccountSearchResponse {
        return accountOperator.search(actor, request, pageable)
    }

    fun setupAccounts() {
        return accountOperator.setup()
    }

    fun searchCategories(actor: Account, request: CategorySearchRequest, pageable: Pageable): CategorySearchResponse {
        return categoryOperator.search(actor, request, pageable)
    }

    fun setupCategories() {
        return categoryOperator.setup()
    }

    fun getInfo() = operators.values.map { it.info() }
    fun getInfo(key: String) = operators[key]?.info()

    fun setup(key: String): SearchOperatorInfo? {
        val operator = operators[key] ?: return null
        operator.setup()
        return operator.info()
    }
}